src/coreclr/vm/jitinterface.cpp

```diff
@@ -10623,10 +10631,37 @@ void* CEEJitInfo::getHelperFtn(CorInfoHelpFunc    ftnNum,         /* IN  */
             dynamicFtnNum == DYNAMIC_CORINFO_HELP_CHKCASTINTERFACE ||
             dynamicFtnNum == DYNAMIC_CORINFO_HELP_CHKCASTCLASS ||
             dynamicFtnNum == DYNAMIC_CORINFO_HELP_CHKCASTCLASS_SPECIAL ||
-            dynamicFtnNum == DYNAMIC_CORINFO_HELP_UNBOX)
+            dynamicFtnNum == DYNAMIC_CORINFO_HELP_UNBOX ||
+            dynamicFtnNum == DYNAMIC_CORINFO_HELP_ARRADDR_ST ||
+            dynamicFtnNum == DYNAMIC_CORINFO_HELP_LDELEMA_REF)
         {
             Precode* pPrecode = Precode::GetPrecodeFromEntryPoint((PCODE)hlpDynamicFuncTable[dynamicFtnNum].pfnHelper);
             _ASSERTE(pPrecode->GetType() == PRECODE_FIXUP);
+            
+            MethodDesc* helperMD = pPrecode->GetMethodDesc();
+            _ASSERT(helperMD != nullptr);
+
+            // Check if the target MethodDesc is already jitted to its final Tier
+            // so we no longer need to use indirections and can emit a direct call instead.
+            NativeCodeVersion::OptimizationTier tier;
+
+            {
+                CodeVersionManager::LockHolder codeVersioningLockHolder;
```

Comment:
This lock is going to be taken when JITing all tier-0 code in the foreground. Can we avoid taking it for the tier-0 foreground JITing?

Reply:
The best I found is `TieredCompilationManager::s_isBackgroundWorkerRunning`, but we might want to do this for foreground OSR as well.

Another problem with this PR that ideally we should slightly reduce call counting treshold for them, otherwise we might see regression in benchmark (when our Tier1 method we bench won't pick up Tier1 helper call because it's not yet ready since it's scheduled to compile after the current method) but that is tricky, not sure how to detect that MethodDesc* is a helper call cheaply, sadly, all bits in MethodDesc flag fields are already occupied.

Reply:
> sure how to detect that MethodDesc* is a helper call cheaply

You can check that it has CastHelpers methodtable.